bot/rules/mentions.py

```diff
@@ -1,24 +1,77 @@
 from typing import Dict, Iterable, List, Optional, Tuple
 
-from discord import Member, Message
+from discord import DeletedReferencedMessage, Member, Message, MessageType, NotFound
+
+import bot
+from bot.log import get_logger
+
+log = get_logger(__name__)
 
 
 async def apply(
     last_message: Message, recent_messages: List[Message], config: Dict[str, int]
 ) -> Optional[Tuple[str, Iterable[Member], Iterable[Message]]]:
-    """Detects total mentions exceeding the limit sent by a single user."""
+    """
+    Detects total mentions exceeding the limit sent by a single user.
+
+    Excludes mentions that are bots, themselves, or replied users.
+
+    In very rare cases, may not be able to determine a
+    mention was to a reply, in which case it is not ignored.
+    """
     relevant_messages = tuple(
         msg
         for msg in recent_messages
         if msg.author == last_message.author
     )
-
+    # We use `msg.mentions` here as that is supplied by the api itself, to determine who was mentioned.
+    # Additionally, `msg.mentions` includes the user replied to, even if the mention doesn't occur in the body.
+    # In order to exclude users who are mentioned as a reply, we check if the msg has a reference
+    #
+    # While we could use regex to parse the message content, and get a list of
+    # the mentions, that solution is very prone to breaking.
+    # We would need to deal with codeblocks, escaping markdown, and any discrepancies between
+    # our implementation and discord's markdown parser which would cause false positives or false negatives.
     total_recent_mentions = sum(
-        not user.bot
+        not (user.bot or msg.author == user)
         for msg in relevant_messages
         for user in msg.mentions
     )
 
+    # No reason to run processing and fetch messages if there are no mentions.
+    if not total_recent_mentions:
+        return None
+
+    # We don't want to include mentions that are to the replied user in message replies.
+    for msg in relevant_messages:
+
+        if msg.type != MessageType.reply and not msg.reference:
```

Comment:
I wasn't sure myself so asked in the dpy server.

Turns out the difference is that `msg.reference` exists for both replies *and* the system message you get from pinning a message.

`MessageType.reply` will only be for replies, and *not* the system message from pins.

TL;DR; This can be `if msg.type != MessageType.reply` since we don't care about the pin system message.